ops/pebble.py

```diff
@@ -134,7 +134,7 @@
     'CheckDict',
     {
         'override': str,
-        'level': Union['CheckLevel', str],
+        'level': str,
```

Comment:
The charms define `CheckDict`s (as part of `Layer`s), right? So this isn't something that's just coming back from Pebble.

So if there's charm code like:

```python
{
...
    "checks": {
        "override": True,
        "level": pebble.CheckLevel.READY,
        ...
    },
...
}
```

then wouldn't this cause problems? I think it's pretty common to be building layers from dictionaries rather than using `Layer` and `Check` and `Service` classes (etc) explicitly.

Reply:
You're right, this seems problematic, I was assuming these dictionaries would only come from pebble, but I notice that `CheckDict` is a public type. @benhoyt, another case where we should keep the `Union` to avoid breaking changes?

Reply:
Hmm, interesting. Would Tony's code actually work, or would the `to_dict` then yield `"level": "CheckLevel.READY"` or something. Yeah, looks like it:

```
>>> print(yaml.dump(d))
exec:
  command: echo foo
http:
  url: https://example.com/
level: !!python/object/apply:ops.pebble.CheckLevel
- ready
override: replace
period: 10s
tcp:
  port: 80
threshold: 5
timeout: 3s
```

So I think this is a plain bug right now (in the `CheckDict` type annotation). Would you agree @tonyandrewmeyer?

Reply:
For all of the changes proposed here, if it currently doesn't work, then I'm fine with the theoretical backwards compatibility break, since in practice it never could have been used, and you could argue that the bug was suggesting it could be.

Reply:
As far as I can tell, both previously and currently, you can pass in a `CheckLevel` member instead of a string, and it will just work. So the annotation change that Tony is highlighting is just making it a type checking error to use an enum member in the dict instead of a `str`. If this is the case, then we should probably allow the `Union` here, right?

Both previously and currently, we essentially do `CheckLevel(check_dict['level'])` in the `Check` constructor, and (e.g.) `CheckLevel(CheckLevel.READY)` just returns `CheckLevel.READY`.

`CheckLevel.to_dict` always expected that `Check.level` could be a `CheckLevel` member, and used `.value` in that case.

How can I reproduce your failing example, @benhoyt?

Reply:
I did it like this:

```
>>> d = {
...             'override': 'replace',
...             'level': 'alive',
...             'period': '10s',
...             'timeout': '3s',
...             'threshold': 5,
...             # Not valid for Pebble to have more than one of http, tcp, and exec,
...             # but it makes things simpler for the unit tests.
...             'http': {'url': 'https://example.com/'},
...             'tcp': {'port': 80},
...             'exec': {'command': 'echo foo'},
...         }
>>> 
>>> d['level'] = pebble.CheckLevel.READY
>>> import yaml
>>> print(yaml.dump(d))
exec:
  command: echo foo
http:
  url: https://example.com/
level: !!python/object/apply:ops.pebble.CheckLevel
- ready
override: replace
period: 10s
tcp:
  port: 80
threshold: 5
timeout: 3s
```

Which I presume is what happens when you create a `pebble.Layer` and then pass it to `add_layer`: it calls `layer.to_yaml()`. And `to_yaml` does `yaml.safe_dump(self.to_dict())`.

Reply:
Thanks, @benhoyt.

Looks like `Pebble.Client.add_layer` does go through `Layer` if you pass a `dict`:
```python
elif isinstance(layer, dict):
    layer_yaml = Layer(layer).to_yaml()
```
Where `Layer`'s constructor then passes any `CheckDict`s through `Check`:
```python
self.checks = {name: Check(name, check) for name, check in d.get('checks', {}).items()}
```
And `Check` now (and previously on main) does the equivalent of:
```python
self.level = CheckLevel(check_dict['level'])  # actually using get + try..except for default/unknown
```
Finally, `Layer.to_yaml` calls `Check.to_dict`, which uses `self.level.value`, and we get a `str` in our yaml the end.

So it seems that it's always worked if `check_dict['level']` is a `CheckLevel` member or a `str`, and it continues to work at runtime, just with a type checking error now. I think that `CheckDict.level` should continue to be typed as `Union[CheckLevel, str]` to remove that type checking error for backwards compatibility.

This doesn't really seem like a bad pattern that we should try to steer charm users away from either, does it?
```python
{
    ...
    "checks": {
        "override": True,
        "level": pebble.CheckLevel.READY,
        ...
    },
    ...
}
```
I kind of prefer it to using strings, as my editor will complain about typos etc. Though you know, if we wanted to provide the same benefits to people using strings, we could consider eventually using `Literal`s instead of `str`s in the Pebble `TypedDict`s, but that's probably something to consider in a future issue.

Reply:
Sounds good. From a quick look at the charms we use for super-tox, no one seems to be using the `"level": pebble.CheckLevel.READY` style.
```bash
ag 'CheckLevel'
```
And using strings seems to be very common.
```bash
ag --case-sensitive '[\'"]level[\'"]:'
```

Reply:
Decision after daily discussion is to go ahead with current approach: basically consider that passing `"level": pebble.CheckLevel.READY,` was a bug -- as the intent was these dicts should all be basic JSON/YAML-ifiable types.

Reply:
Note: Following a question in daily discussion, using (e.g.) `pebble.CheckLevel.READY` works at runtime, but not because of any explicit special casing of `str` vs `CheckLevel`.

It only works because of how enum constructors are happy to take either a value or a member as their argument (i.e. `CheckLevel('ready') == CheckLevel.READY == CheckLevel(CheckLevel.READY)`.

As decided, `level": pebble.CheckLevel.READY` will result in a type checking error, but we're not going to add a runtime error (just like we decided not to in other cases).